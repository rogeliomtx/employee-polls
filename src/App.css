/* Auth */
body {
  background-color: #f5f5f5;
}

.card.auth {
  width: 18rem;
}

img.avatar, div.avatar {
  width: 100px;
  height: 100px;
  display: block;
  margin: 0 auto;
  border-radius: 100%;
}

div.avatar {
  background: #0d6efd;
  color: #fff;
  border-radius: 100%;
}

div.avatar span {
  font-size: 3em;
  line-height: 100px;
  display: block;
  text-align: center;
}

.navbar .avatar {
  width: 30px;
  height: 30px;
  display: inline-block;
  margin-right: .5em;
}

.leaderboard .avatar {
  width: 50px;
  height: 50px;
  display: inline-block;
  margin-right: .5em;
}

.navbar .logout {
  cursor: pointer;
}