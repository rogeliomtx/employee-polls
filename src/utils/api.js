import { _getUsers, _getQuestions, _saveQuestion, _saveQuestionAnswer } from './_DATA';

export async function getInitialData () {
  return Promise.all([
    _getUsers(),
    _getQuestions(),
  ]).then(([users, questions]) => ({
    users,
    questions
  }))
}

export async function saveQuestion (question) {
  return _saveQuestion(question)
}

export async function saveQuestionAnswer ({ authedUser, question, answer }) {
  return _saveQuestionAnswer({
    authedUser,
    qid: question.id,
    answer
  }).then((saved) => {
    return { authedUser, answer, questionId: question.id }
  })
}

export async function login (id, password) {
  return _getUsers().then((users) => {
    const user = Object.values(users).find((user) => user.id === id && user.password === password);
    return user;
  })
}
