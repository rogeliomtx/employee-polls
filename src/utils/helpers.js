export function formatQuestion (question, author, authedUser) {
  const { id, optionOne, optionTwo, timestamp } = question

  return {
    id,
    timestamp,
    optionOne,
    optionTwo,
    author,
    hasAnswered: optionOne.votes.includes(authedUser) || optionTwo.votes.includes(authedUser)
  }
}