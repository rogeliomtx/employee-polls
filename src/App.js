import { Route, Routes } from "react-router-dom";
import { useEffect, Fragment } from "react";
import { connect } from "react-redux";
import { LoadingBar } from "react-redux-loading-bar";

import "bootstrap/dist/css/bootstrap.min.css";
import './App.css';

import Home from './components/Home';
import Leaderboard from './components/Leaderboard';
import Login from './components/Login';
import CreateQuestion from './components/CreateQuestion';
import QuestionDetails from './components/QuestionDetails';
import Navbar from './components/Navbar';
import NotFound from './components/NotFound';
import ProtectedRoute from "./components/ProtectedRoute";

import { handleInitialData } from "./actions/shared";

function App({ dispatch, loading }) {
  useEffect(() => {
    dispatch(handleInitialData());
  });

  return (
    <Fragment>
      <LoadingBar />
      {
        loading ? <div>Loading...</div> : (
          <div>
            <Navbar />
            <div className="container">
              <Routes>                
                <Route path="/" element={
                  <ProtectedRoute>
                    <Home />
                  </ProtectedRoute>
                } />
                <Route path="/login" element={<Login />} />
                <Route path="/leaderboard" element={
                  <ProtectedRoute>
                    <Leaderboard />
                  </ProtectedRoute>
                } />
                <Route path="/add" element={
                  <ProtectedRoute>
                    <CreateQuestion />
                  </ProtectedRoute>
                } />
                <Route path="/questions/:questionId" element={
                  <ProtectedRoute>
                    <QuestionDetails />
                  </ProtectedRoute>
                } />
                <Route path="/*" element={<NotFound />} />
              </Routes>
            </div>
          </div>
        )
      }
    </Fragment>
  );
}

const mapStateToProps = ({ users, authedUser }) => {
  return {
    loading: users === null,  // if users are loaded, loading is false
    authedUser,
  }
}
export default connect(mapStateToProps)(App);
