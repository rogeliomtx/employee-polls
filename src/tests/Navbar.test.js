import { MemoryRouter } from "react-router";
import { screen, within } from '@testing-library/react';

import { renderWithProviders, getMockState } from "./utils";
import App from "../App";

const authedUserState = {preloadedState: getMockState("sarahedo")};
const defaultState = {preloadedState: getMockState()};

const component = (
  <MemoryRouter>
    <App />
  </MemoryRouter>
);

describe("Navbar with authed user", () => {
  it("should match the snapshot", () => {
    const view = renderWithProviders(component, authedUserState);
    expect(view.container).toMatchSnapshot();
  });

  it("should render the menu when user is logged in", () => {
    // -> Test that the navbar renders correctly when the user is logged in.
    // -> Check that all links are present
    renderWithProviders(component, authedUserState);
    expect(screen.getByText(/home/i)).toBeInTheDocument();
    expect(screen.getByText(/Leaderboard/i)).toBeInTheDocument();
    expect(screen.getByText(/Add Question/i)).toBeInTheDocument();
    expect(screen.getByText(/Logout/i)).toBeInTheDocument();
  });

  it("should render the user's name and avatar", () => {
    // -> Test that the navbar renders correctly when the user is logged in.
    // -> Check that the user's name and avatar is displayed.
    renderWithProviders(component, authedUserState);
    const navigation = screen.getByRole('navigation');
    
    // user.name
    expect(within(navigation).getByText(/sarah edo/i)).toBeInTheDocument();
    // user.avatar
    expect(within(navigation).getByRole('img', {name: /avatar/i})).toBeInTheDocument();

    expect(screen.getByText(/logout/i)).toBeInTheDocument();
  });

  it("should not render login", () => {
    renderWithProviders(component, authedUserState);
    expect(screen.queryByText(/login/i)).not.toBeInTheDocument();
  });
});

describe("Navbar without authed user", () => {
  it("should match the snapshot", () => {
    const view = renderWithProviders(component, defaultState);
    expect(view.container).toMatchSnapshot();
  });
  
  it("should render the menu when user is not logged in", () => {
    renderWithProviders(component, defaultState);
    expect(screen.getByText(/home/i)).toBeInTheDocument();
    expect(screen.getByText(/Leaderboard/i)).toBeInTheDocument();
    expect(screen.getByText(/Add Question/i)).toBeInTheDocument();
    expect(screen.getByRole('link', {name: /login/i})).toBeInTheDocument();
  });

  it("shouldn't render the user's name and avatar", () => {
    renderWithProviders(component, defaultState);
    
    const navigation = screen.getByRole('navigation');

    // user.name
    expect(within(navigation).queryByText(/sarah edo/i)).not.toBeInTheDocument();
    // user.avatar
    expect(within(navigation).queryByRole('img', {name: /avatar/i})).not.toBeInTheDocument();
  });

  it("should not render logout", () => {
    renderWithProviders(component, defaultState);
    expect(screen.queryByText(/logout/i)).not.toBeInTheDocument();
  });
});