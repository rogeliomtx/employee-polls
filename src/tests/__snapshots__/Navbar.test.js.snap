// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar with authed user should match the snapshot 1`] = `
<div>
  <div />
  <div>
    <nav
      class="navbar navbar-expand-lg bg-primary"
      data-testid="menu"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand link-light"
          href="/"
        >
          Employee Polls
        </a>
        <div
          class="collapse navbar-collapse"
          id="navbarText"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link active link-light"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/add"
              >
                Add question
              </a>
            </li>
          </ul>
          <ul
            class="navbar-nav mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="Avatar"
                    class="avatar"
                    src="/assets/avatar_1.jpg"
                  />
                </div>
                Sarah Edo
              </span>
            </li>
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light logout"
              >
                Logout
              </span>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        role="contentinfo"
      >
        <div
          class="card mt-5 p-3"
        >
          <ul
            class="nav nav-pills mt-3"
            role="tablist"
          >
            <li
              class="nav-item"
              role="presentation"
            >
              <button
                class="nav-link active"
                id="unanswered-questions"
                role="tab"
                type="button"
              >
                Unanswered Questions
              </button>
            </li>
            <li
              class="nav-item"
              role="presentation"
            >
              <button
                class="nav-link "
                id="answered-questions"
                role="tab"
                type="button"
              >
                Answered Questions
              </button>
            </li>
          </ul>
          <div
            class="tab-content"
          >
            <div
              class="tab-pane fade mt-5 active show"
              id="unanswered-questions"
              role="tabpanel"
            >
              <div
                class="row row-cols-4"
                data-testid="unanswered-questions"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Navbar without authed user should match the snapshot 1`] = `
<div>
  <div />
  <div>
    <nav
      class="navbar navbar-expand-lg bg-primary"
      data-testid="menu"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand link-light"
          href="/"
        >
          Employee Polls
        </a>
        <div
          class="collapse navbar-collapse"
          id="navbarText"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link active link-light"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/add"
              >
                Add question
              </a>
            </li>
          </ul>
          <ul
            class="navbar-nav mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/login"
              >
                Login
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="card auth mt-5 m-auto"
        role="contentinfo"
      >
        <div
          class="card-header bg-primary text-light"
        >
          <h2>
            Employee Polls
          </h2>
        </div>
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Login
          </h5>
          <form>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="on"
                class="form-control"
                placeholder="Enter your username"
                type="text"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                placeholder="Enter your password"
                type="password"
              />
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Sign in
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
